### DISCUSSION TITLE
Any way to pass UE4 comments to Hover/Completion with UE4's reflection system?

### DISCUSSION THREAD

**Discussion by boocs** [microsoft/vscode-cpptools#7248] [Q&A] [answered]
```
        /**
	 * Sets whether the component is active or not
	 * @param bNewActive - The new active state of the component
	 * @param bReset - Whether the activation should happen even if ShouldActivate returns false.
	 */
	UFUNCTION(BlueprintCallable, Category="Components|Activation", meta=(UnsafeDuringActorConstruction="true"))
	virtual void SetActive(bool bNewActive, bool bReset=false);
```

The UFUNCTION does nothing in code but it does get read by UE4 to build Blueprint functionality. 

These Macros block the documentation from appearing on hover and completion though. I tried some #ifdef __INTELLISENSE__ tricks with no luck. MS got rid of the macro trick that could create a comment.

Any cool cpp.hint tricks that could do this?

Replacing the macro with a blank line won't work. (edit: The macros actually are a blank line)

Maybe a cpptools fix? A regex for lines that will get ignored by hover/completion?

**Comment by sean-mcmanus** [comment, marked as answer]
Yeah, you can create a cpp.hint with 
```
#define UFUNCTION(x, y, z) extern
```
Let us know if that doesn't work. The main thing is that the cpp.hint definition can't be non-empty.

What does "MS got rid of the macro trick that could create a comment." mean?

**Comment by boocs** [comment]
@sean-mcmanus 

(edited) It works! Both cpp.hint files in root workspace directory.

cpp.hint:
```
#define UFUNCTION(...) extern
#define UPROPERTY(...) extern
```
Currently only doing the two but could be tested on other UMacros.

---
I meant using ## to  replace a macro with a comment. I found one that worked though! Didn't work in the past with ones I tried and they said MS removed support for it. Weird that it's this simple:
```
#define COMMENT / ## /
```

**Comment by boocs** [comment]
@sean-mcmanus 

Thanks again for the help! I had tried fixing that multiple times. 

Do you think other compilers will be compatible with the fix? Especially clang?

**Reply by sean-mcmanus** [reply]
The cpp.hint workaround operates in "tag parser" mode, which should work for all compilers (IntelliSense modes).

**Comment by boocs** [comment]
@sean-mcmanus 

I forgot to mention anomalies during testing. This could be my fault but something to check. 

1. Testing cpp.hint seemed like it wasn't working
2. I used Process Monitor and nothing was looking for/reading any cpp.hint files when VSCode launches
3. Only when I modified a file(header/source) did VSCode start looking for/reading cpp.hint files
4. It seemed to started working after that

1. When it started working I reloaded VSCode
2. cpp.hint still isn't being read/searched for on startup
3. But... cpp.hint still works!
4. So I'm guessing whatever cpp.hint does gets cached somewhere?

This could be an anomaly with Process Monitor or just how VSCode works. Of course there's a good chance of user error.

**Reply by sean-mcmanus** [reply]
The cpp.hint will affect the tag parsing, which has its results stored to the database file. It looks like we have a bug in regards to cpp.hint changes not automatically causing the IntelliSense database to reset. See https://github.com/microsoft/vscode-cpptools/issues/5467 . So the current workaround is you may need to do a Reset IntelliSense Database command after modifying cpp.hint.
